wasm/wat/parser.go

```diff
@@ -0,0 +1,248 @@
+package wat
+
+import (
+	"errors"
+	"fmt"
+)
+
+type currentField byte
+
+const (
+	// fieldInitial is the first position in the source being parsed.
+	fieldInitial currentField = iota
+	fieldModule
+	fieldModuleImport
+	fieldModuleImportFunc
+	fieldModuleStart
+)
+
+type ModuleParser struct {
+	source []byte
+	module *module
+	// currentStringCount allows us to unquote the _import.module and _import.name fields, differentiating empty
+	// from never set, without making _import.module and _import.name pointers.
+	currentStringCount int
+	// currentField is the parser and error context.
+	// This is set after reading a field name, ex "module", or after reaching the end of one, ex ')'.
+	currentField
+	// tokenParser is called by lex, and changes based on the currentField.
+	// The initial tokenParser is ensureLParen because %.wat files must begin with a '(' token (ignoring whitespace).
+	tokenParser
+	// currentImportIndex allows us to track the relative position of imports regardless of what they describe
+	currentImportIndex int
+	// currentImportModule is the first string of the current import. Ex. "Math" if (import "Math" "PI" (func))
+	currentImportModule string
+	// currentImportName is the second string of the current import. Ex. "PI" if (import "Math" "PI" (func))
+	currentImportName string
+	// currentImportHasDesc tracks if we found a description (Ex. (func)) of the current import.
+	currentImportHasDesc bool
+}
+
+// parseModule parses the configured source into a module. This function returns when the source is exhausted or an
+// error occurs.
+//
+// Here's a description of the return values:
+// * module is the result of parsing or nil on error
+// * err is a formatError invoking the parser, dangling block comments or unexpected characters.
+func parseModule(source []byte) (*module, error) {
+	p := ModuleParser{source: source, module: &module{}}
+
+	// A valid source must begin with the token '(', but it could be preceded by whitespace or comments. For this
+	// reason, we cannot enforce source[0] == '(', and instead need to start the lexer to check the first token.
+	p.tokenParser = p.ensureLParen
+	line, col, err := lex(p.parse, p.source)
+	if err != nil {
+		return nil, &formatError{line, col, p.errorContext(), err}
+	}
+	return p.module, nil
+}
+
+// parse calls the delegate ModuleParser.tokenParser
+func (p *ModuleParser) parse(tok tokenType, tokenBytes []byte, line, col int) error {
+	return p.tokenParser(tok, tokenBytes, line, col)
+}
+
+func (p *ModuleParser) ensureLParen(tok tokenType, _ []byte, _, _ int) error {
+	if tok != tokenLParen {
+		return fmt.Errorf("expected '(', but found %s", tok)
+	}
+	p.tokenParser = p.startField
+	return nil
+}
+
+func (p *ModuleParser) startField(tok tokenType, tokenBytes []byte, _, _ int) (err error) {
```

Comment:
wanna know that `Field` is a term in spec (for text format) or specific in this code?

Reply:
field is used a lot in the spec indeed. incidentally, I learned that after looking at Wabt, then verified also as I thought it kind strange term. it isn't consistent but "A module consists of a sequence of fields that can occur in any order." https://www.w3.org/TR/wasm-core-1/#modules%E2%91%A0%E2%91%A2

Reply:
of course they say "any order" then later say imports must be first 🤷

Reply:
🤷 🧐